### PR TITLE
TextDecoder: streaming decode, ignoreBOM

### LINKED ISSUE #5600: Implement TextDecoder#ignoreBOM.
Blocks #5601.
- Spec: https://encoding.spec.whatwg.org/#dom-textdecoder-ignorebom.
- Code: `components/script/dom/textdecoder.rs`, `components/script/dom/webidls/TextDecoder.webidl`.
- Test: `/encoding/textdecoder-ignorebom.html`, `/encoding/idlharness.html`.

### LINKED ISSUE #13234: Implement streaming support for TextDecoder
Blocks #5601.
- Spec: https://encoding.spec.whatwg.org/#dom-textdecoder-decode.
- Code: `components/script/dom/textdecoder.rs`, `components/script/dom/webidls/TextDecoder.webidl`.
- Test: `/encoding/textdecoder-fatal-streaming.html`, `/encoding/textdecoder-streaming.html`.

### PR DESCRIPTION
**Pull request by talklittle**
Implement streaming decode and ignoreBOM flag for TextDecoder.

https://encoding.spec.whatwg.org/#dom-textdecoder-decode

https://encoding.spec.whatwg.org/#dom-textdecoder-ignorebom

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #13234 (github issue number if applicable).
- [x] These changes fix #5600 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [x] These changes do not require tests because the wpt tests are used for testing:

* /encoding/textdecoder-fatal-streaming.html
* /encoding/textdecoder-streaming.html
* /encoding/textdecoder-ignorebom.html

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20431)

### CODE REVIEW

**Comment by talklittle** [comment]
This PR also fixes the tests `/encoding/textdecoder-byte-order-marks.html`. Unsure if that means issue #13233 is considered fixed as well?

File: components/script/dom/textdecoder.rs [hunk 1]
```diff
@@ -11,31 +11,44 @@ use dom::bindings::root::DomRoot;
...
 #[dom_struct]
 pub struct TextDecoder {
     reflector_: Reflector,
     encoding: &'static Encoding,
     fatal: bool,
+    ignoreBOM: bool,
+    #[ignore_malloc_size_of = "defined in encoding_rs"]
+    decoder_: RefCell<Decoder>,
+    in_stream_: RefCell<Vec<u8>>,
+    do_not_flush_: Cell<bool>,
```
Comment (jdm) [inline components/script/dom/textdecoder.rs]: nit: no need for the trailing _s

File: components/script/dom/textdecoder.rs [hunk 2]
```diff
@@ -65,30 +78,60 @@ impl TextDecoderMethods for TextDecoder {
...
-                    match self.encoding.decode_without_bom_handling_and_without_replacement(&vec) {
-                        Some(s) => s,
-                        None => return Err(Error::Type("Decoding failed".to_owned())),
-                    }
-                } else {
-                    let (s, _has_errors) = self.encoding.decode_without_bom_handling(&vec);
-                    s
-                };
-                Ok(USVString(s.into_owned()))
+        if !self.do_not_flush_.get() {
```
Comment (jdm) [inline components/script/dom/textdecoder.rs]: Let's add a `// Step 1` comment here (and similar for the rest of this method).

**Review by jdm — CHANGES_REQUESTED** [review]
This looks really great! Thanks for working on this!

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
Hooray, lots of other passing tests!

**Pull request merged**